build/build/src/ide/web/download.rs

```diff
@@ -0,0 +1,105 @@
+//! Support for downloading files.
+
+use crate::prelude::*;
+
+use ide_ci::cache::Cache;
+use ide_ci::cache::Storable;
+use reqwest::IntoUrl;
+use std::fs::File;
+
+
+
+// ===========================
+// === Download with cache ===
+// ===========================
+
+/// Check whether the specified URL has been downloaded and cached. If not, download it using the
+/// specified downloader, and store it in the cache. Opens the file in the cache and returns it.
+///
+/// The `filename` parameter will be used to name the file within its directory in the cache; the
+/// name does not matter to application operation, but only serves to inform anyone inspecting the
+/// cache directory.
+pub async fn get_file_from_cache_or_download(
+    filename: impl AsRef<Path>,
+    cache: &Cache,
+    downloader: impl Downloader,
+    url: impl IntoUrl,
+) -> Result<File> {
+    let url = url.into_url()?;
+    let filename = filename.as_ref().into();
+    let downloader = Arc::new(downloader);
+    let cache_path = cache.get(FixedContentUrl { url, filename, downloader }).await?;
+    Ok(File::open(&cache_path)?)
```

Comment:
Prefer `ide_ci::fs::open` for better diagnostics.
Also, if this function is asynchronous, I'd also consider using tokio's file wrapper.